### DISCUSSION TITLE
Performance analysis

### DISCUSSION THREAD

**Discussion by letto4135** [fastapi/fastapi#3792] [Questions] [answered]
Node/express vs fastapi with uvicorn and django with uvicorn.

Not that I went into this too deep, but I ran a few tests on node vs fastapi and django and am sharing the results.
Tests counted all responses over 10 seconds 10 times and averaged the times.

Overall node won, on simple hello world app fastapi beat out node, but as soon as more data was introduced (5 real estate listings of data from https://api.listhub.com/api/swagger/#/odata/property) fastapi quickly fell behind just returning the raw data. Django was able to process more requests than fastapi as well. All the requests were is returning a small bit of json data, not pulling it from that API or anything.

Basic hello world requests processed
2649.9 node
2690.9 fastapi

Requests processed with real estate data
2163.5 node
1188.4 fastapi
1741.1 django

I don't think that fastapi was supposed to be better than node or anything, but I would have thought it would have been faster than django at least. Anyone want to help confirm or deny the findings?

**Comment by HarrySky** [comment]
Maybe share the benchmarking code first?

FastAPI is fast when I/O is used (with async of course), when you just return JSON data synchronously - it will be slow. But when you pull the data with asycpg, concurrently and running on uvloop - I bet it will beat Django. Not sure about Node

**Reply by letto4135** [reply]
I literally used exactly what the site shows for "first steps" and just replace what it was returning with a larger json response and ran it exactly how it shows with "uvicorn main:app --reload"


```
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}
```

**Reply by letto4135** [reply]
Also, express was just the simple helloworld code from their site as well, just with the response replaced with res.json and the same data as fastapi.
Django was similar, just basic view that returned a JsonResponse.

```
const express = require('express')
const app = express()
const port = 3000

app.get('/', (req, res) => {
  res.send('Hello World!')
})

app.listen(port, () => {
  console.log(`Example app listening at http://localhost:${port}`)
})
```

**Comment by tiangolo** [comment, marked as answer]
There's a third party, TechEmpower, running quite complete and professional benchmarks, you can check the code for each one there to see tricks to squeeze performance.

https://www.techempower.com/benchmarks/

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
